File: ios/App/Plugins/SessionExpiredListenerPlugin.swift
import Foundation
import Capacitor

/**
 * SessionExpiredListenerPlugin para iOS
 * Escuta notificações nativas de sessão expirada (HTTP 401)
 * Notifica o JavaScript para renovar o token
 * Equivalente ao BroadcastReceiver do Android
 */
@objc(SessionExpiredListenerPlugin)
public class SessionExpiredListenerPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "SessionExpiredListenerPlugin"
    public let jsName = "SessionExpiredListener"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "echo", returnType: CAPPluginReturnPromise)
    ]
    
    private static let sessionExpiredNotification = Notification.Name("tech.orizon.ampara.SESSION_EXPIRED")
    
    public override func load() {
        super.load()
        registerSessionExpiredObserver()
    }
    
    deinit {
        unregisterSessionExpiredObserver()
    }
    
    private func registerSessionExpiredObserver() {
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(handleSessionExpired(_:)),
            name: SessionExpiredListenerPlugin.sessionExpiredNotification,
            object: nil
        )
        CAPLog.print("✅ Session expired observer registered")
    }
    
    private func unregisterSessionExpiredObserver() {
        NotificationCenter.default.removeObserver(
            self,
            name: SessionExpiredListenerPlugin.sessionExpiredNotification,
            object: nil
        )
        CAPLog.print("Session expired observer unregistered")
    }
    
    @objc private func handleSessionExpired(_ notification: Notification) {
        let source = notification.userInfo?["source"] as? String ?? "unknown"
        CAPLog.print("📡 Session expired notification received from: \(source)")
        
        notifyListeners("sessionExpired", data: ["source": source])
    }
    
    @objc func echo(_ call: CAPPluginCall) {
        let value = call.getString("value") ?? ""
        call.resolve(["value": value])
    }
    
    // MARK: - Public API for Native Services
    
    /**
     * Método estático para ser chamado por serviços nativos (KeepAlive, etc)
     * Envia notificação que será capturada pelo observer
     */
    public static func notifySessionExpired(source: String) {
        CAPLog.print("🔔 Posting session expired notification from: \(source)")
        NotificationCenter.default.post(
            name: sessionExpiredNotification,
            object: nil,
            userInfo: ["source": source]
        )
    }
}

File: android/app/src/main/java/tech/orizon/ampara/plugins/SessionExpiredListenerPlugin.java
package tech.orizon.ampara.plugins;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.util.Log;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

/**
 * Plugin para escutar broadcasts nativos de sessão expirada
 * Permite que o JavaScript seja notificado quando o ping nativo detecta HTTP 401
 */
@CapacitorPlugin(name = "SessionExpiredListener")
public class SessionExpiredListenerPlugin extends Plugin {
    
    private static final String TAG = "SessionExpiredListener";
    private BroadcastReceiver sessionExpiredReceiver;
    
    @Override
    public void load() {
        super.load();
        registerSessionExpiredReceiver();
    }
    
    @Override
    protected void handleOnDestroy() {
        super.handleOnDestroy();
        unregisterSessionExpiredReceiver();
    }
    
    private void registerSessionExpiredReceiver() {
        sessionExpiredReceiver = new BroadcastReceiver() {
            @Override
            public void onReceive(Context context, Intent intent) {
                Log.d(TAG, "Session expired broadcast received from native");
                
                // Notificar o JavaScript via evento
                JSObject ret = new JSObject();
                ret.put("source", intent.getStringExtra("source"));
                notifyListeners("sessionExpired", ret);
            }
        };
        
        IntentFilter filter = new IntentFilter("tech.orizon.ampara.SESSION_EXPIRED");
        getContext().registerReceiver(sessionExpiredReceiver, filter, Context.RECEIVER_NOT_EXPORTED);
        Log.d(TAG, "Session expired receiver registered");
    }
    
    private void unregisterSessionExpiredReceiver() {
        if (sessionExpiredReceiver != null) {
            try {
                getContext().unregisterReceiver(sessionExpiredReceiver);
                sessionExpiredReceiver = null;
                Log.d(TAG, "Session expired receiver unregistered");
            } catch (IllegalArgumentException e) {
                // Receiver já foi desregistrado
                Log.w(TAG, "Receiver was already unregistered");
            }
        }
    }
    
    @PluginMethod
    public void echo(PluginCall call) {
        String value = call.getString("value");
        JSObject ret = new JSObject();
        ret.put("value", value);
        call.resolve(ret);
    }
    
    /**
     * Método público para ser chamado diretamente pelo KeepAliveService
     * Notifica o JavaScript sobre sessão expirada
     */
    public void notifySessionExpiredFromService(String source) {
        Log.d(TAG, "Session expired notification received from service: " + source);
        JSObject ret = new JSObject();
        ret.put("source", source);
        notifyListeners("sessionExpired", ret);
        Log.d(TAG, "Session expired event sent to JavaScript");
    }
}
